include/Target/Btor/BtorToBtorIRTranslation.h

```diff
@@ -0,0 +1,190 @@
+        modelFile = fopen(s.c_str(), "r");
+    }
+
+  ~Deserialize() {
+      if (model) {
+        btor2parser_delete(model);
+      }
+      if (modelFile) {
+        fclose(modelFile);
+      }
+  }
+
+///===----------------------------------------------------------------------===//
+/// Parse btor2 file
+///===----------------------------------------------------------------------===//
+
+  std::vector<Btor2Line *> inputs;
+  std::vector<Btor2Line *> states;
+  std::vector<Btor2Line *> bads;
+  std::vector<Btor2Line *> inits;
+  std::vector<Btor2Line *> nexts;
+  std::vector<Btor2Line *> constraints;
+ 
+  std::map<int64_t, Btor2Line *> reachedLines;
+  
+  bool parseModelIsSuccessful();
+
+///===----------------------------------------------------------------------===//
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+  
+  std::map<int64_t, Value> cache;
+
+  OwningOpRef<FuncOp> buildInitFunction();
+  OwningOpRef<FuncOp> buildNextFunction();
+  
+ private: 
+///===----------------------------------------------------------------------===//
+/// Parse btor2 file
+///===----------------------------------------------------------------------===//
+  
+  Btor2Parser *model = nullptr;
+  FILE *modelFile = nullptr;
+
+  void parseModelLine(Btor2Line *l);
+
+  void filterInits(){
+    size_t i = 0;
+    for (size_t j = 0, sz = inits.size(); j < sz; ++j) {
+        if (inits.at(j)) {
+            inits[i++] = inits.at(j);
+        }
+    }
+    inits.resize(i);
+  }
+
+  void filterNexts() {
+    size_t i = 0;
+    for (size_t j = 0, sz = nexts.size(); j < sz; ++j) {
+      if (nexts.at(j)) {
+        nexts[i++] = nexts.at(j);
+      }
+    }
+    nexts.resize(i);
+  }
+
+///===----------------------------------------------------------------------===//
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+
+  MLIRContext *context;
+  OpBuilder builder;
+  Location unknownLoc;
+  
+  void toOp(Btor2Line *line);
+  bool isValidChild(Btor2Line * line);
+  void createNegateLine(int64_t curAt, Value child);
+  Operation * createMLIR(const Btor2Line *line, const int64_t *kids);
+
+  template <typename btorOp>
+  Operation * buildBinaryOp(Value lhs, Value rhs) {
+    auto res = builder.create<btorOp>(unknownLoc, lhs, rhs);
+    return res;
+  }
+
+  template <typename btorOp>
+  Operation * buildComparisonOp(btor::BtorPredicate pred,
+                                Value lhs, Value rhs) {
+    auto res = builder.create<btorOp>(unknownLoc, pred, lhs, rhs);
+    return res;
+  }
+
+  template <typename btorOp>
+  Operation * buildOverflowOp(Value lhs, Value rhs) {
+    auto res = builder.create<btorOp>(unknownLoc, 
+                                    builder.getIntegerType(1), 
+                                    lhs, rhs);
+    return res;
+  }
+
+  Operation * buildConstantOp(uint32_t width, std::string str, uint32_t radix) {
```

Comment:
`std::string` passed by value, is that intentional?

Reply:
Yes. They are created on the stack, in the client code, so it wouldn't make sense to pass by reference

Reply:
Passing as a `const` reference now :)